### PR TITLE
Migrating existing rest fe tests to real cases

### PR DESCRIPTION
**Pull request by yaooqinn**
### _Why are the changes needed?_

In this PR, we target the existing UTs from a noop server to a real shared Kyuubi server.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiRestFrontendService.scala [hunk 1]
```diff
@@ -76,7 +76,7 @@ class KyuubiRestFrontendService(override val serverable: Serverable)
 
   override def connectionUrl: String = {
     checkInitialized()
-    s"${serverAddr.getCanonicalHostName}:$portNum"
+    s"${serverAddr.getCanonicalHostName}:${connector.getLocalPort}"
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiRestFrontendService.scala:79]: there's another bug here, in this pr, we only fix a bug for port==0

File: kyuubi-server/src/test/scala/org/apache/kyuubi/RestFrontendTestHelper.scala [hunk 2]
```diff
@@ -46,40 +48,29 @@ object RestFrontendTestHelper {
   }
 }
 
-trait RestFrontendTestHelper {
+trait RestFrontendTestHelper extends WithKyuubiServer {
```
Comment (yaooqinn) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/RestFrontendTestHelper.scala:51]: refine to use a real singleton Kyuubi server for testing

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/KyuubiRestFrontendServiceSuite.scala [hunk 3]
```diff
@@ -17,80 +17,39 @@
...
-import org.scalatest.time.SpanSugar._
-
-import org.apache.kyuubi.{KyuubiFunSuite, RestFrontendTestHelper}
-import org.apache.kyuubi.config.KyuubiConf
-import org.apache.kyuubi.service.NoopRestFrontendServer
-import org.apache.kyuubi.service.ServiceState._
-
-class KyuubiRestFrontendServiceSuite extends KyuubiFunSuite with RestFrontendTestHelper {
-
-  test("kyuubi REST frontend service basic") {
```
Comment (yaooqinn) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/KyuubiRestFrontendServiceSuite.scala:33]: this case is not very necessary as most logic will be verified in RestFrontendTestHelper

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/KyuubiRestFrontendServiceSuite.scala [hunk 4]
```diff
@@ -17,80 +17,39 @@
...
+    response = webTarget.path("api/v1/exception").request().get()
+    assert(500 == response.getStatus)
+    assert(response.getStatusInfo.getReasonPhrase.equalsIgnoreCase("server error"))
+  }
 
-      // send a exists request but wrong http method
-      response = webTarget.path("api/v1/ping").request().post(null)
-      assert(405 == response.getStatus)
-      assert(response.getStatusInfo.getReasonPhrase.equalsIgnoreCase("method not allowed"))
+  test("swagger ui") {
```
Comment (yaooqinn) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/KyuubiRestFrontendServiceSuite.scala:46]: add 2 swagger tests

**Comment by yaooqinn** [comment]
cc @pan3793 @yanghua @simon824 @cfmcgrady thanks

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala [hunk 5]
```diff
@@ -34,88 +34,62 @@ import org.apache.kyuubi.session.SessionHandle
 
 class SessionsResourceSuite extends KyuubiFunSuite with RestFrontendTestHelper {
 
-  test("test open and count session") {
+  test("test open/close and count session") {
```
Comment (yanghua) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala:37]: Remove the started `test` in the character?

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala [hunk 6]
```diff
@@ -126,61 +100,57 @@ class SessionsResourceSuite extends KyuubiFunSuite with RestFrontendTestHelper {
...
+      .request(MediaType.APPLICATION_JSON_TYPE)
+      .post(Entity.entity(requestObj, MediaType.APPLICATION_JSON_TYPE))
+
+    // get session list
+    var response2 = webTarget.path("api/v1/sessions").request().get()
+    assert(200 == response2.getStatus)
+    val sessions1 = response2.readEntity(classOf[SessionList])
+    assert(sessions1.sessionList.nonEmpty)
+
+    // close a opened session
```
Comment (yanghua) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala:113]: an?

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala [hunk 7]
```diff
@@ -34,88 +34,62 @@ import org.apache.kyuubi.session.SessionHandle
...
-
-      val sessionHandle = response.readEntity(classOf[SessionHandle])
+    // verify the open session count
+    response = webTarget.path("api/v1/sessions/count").request().get()
+    val openedSessionCount = response.readEntity(classOf[SessionOpenCount])
+    assert(openedSessionCount.openSessionCount == 1)
 
-      assert(sessionHandle.protocol.getValue == 1)
-      assert(sessionHandle.identifier != null)
+    // close a opened session
```
Comment (yanghua) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala:61]: an?

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala [hunk 8]
```diff
@@ -34,88 +34,62 @@ import org.apache.kyuubi.session.SessionHandle
...
-
-    withKyuubiRestServer { (_, _, _, webTarget) =>
-      var response = webTarget.path("api/v1/sessions")
-        .request(MediaType.APPLICATION_JSON_TYPE)
-        .post(Entity.entity(requestObj, MediaType.APPLICATION_JSON_TYPE))
-
-      assert(200 == response.getStatus)
-
-      val sessionHandle = response.readEntity(classOf[SessionHandle])
+    // verify the open session count
```
Comment (yanghua) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala:56]: opened?

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala [hunk 9]
```diff
@@ -34,88 +34,62 @@ import org.apache.kyuubi.session.SessionHandle
...
-        s"${sessionHandle.identifier.secretId}|${sessionHandle.protocol.getValue}"
-      response = webTarget.path(s"api/v1/sessions/$serializedSessionHandle").request().delete()
-      assert(200 == response.getStatus)
-
-      // verify the open session count again
-      response = webTarget.path("api/v1/sessions/count").request().get()
-      val openedSessionCount = response.readEntity(classOf[SessionOpenCount])
-      assert(openedSessionCount.openSessionCount == 0)
-    }
+    // verify the open session count again
```
Comment (yanghua) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala:67]: opened?

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala [hunk 10]
```diff
@@ -17,116 +17,116 @@
...
+      .request(MediaType.APPLICATION_JSON_TYPE)
+      .put(Entity.entity(OpActionRequest("close"), MediaType.APPLICATION_JSON_TYPE))
+    assert(200 == response.getStatus)
+
+    response = webTarget.path(s"api/v1/operations/$opHandleStr/event")
+      .request(MediaType.APPLICATION_JSON_TYPE).get()
+    assert(404 == response.getStatus)
   }
 
   test("test get result set metadata") {
```
Comment (yanghua) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala:82]: remove the started `test`?

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/KyuubiRestFrontendServiceSuite.scala [hunk 11]
```diff
@@ -17,80 +17,39 @@
...
+    var response = webTarget.path("api/v1/pong").request().get()
+    assert(404 == response.getStatus)
+    assert(response.getStatusInfo.getReasonPhrase.equalsIgnoreCase("not found"))
+
+    // send a exists request but wrong http method
+    response = webTarget.path("api/v1/ping").request().post(null)
+    assert(405 == response.getStatus)
+    assert(response.getStatusInfo.getReasonPhrase.equalsIgnoreCase("method not allowed"))
+
+    // send a request but throws a exception on the server side
```
Comment (yanghua) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/KyuubiRestFrontendServiceSuite.scala:40]: an exception?

**Review by yanghua — COMMENTED** [review]
Left some comments.

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/KyuubiRestFrontendServiceSuite.scala [hunk 12]
```diff
@@ -17,80 +17,39 @@
...
-          val s = Source.fromURL(s"http://$host:$port/api/v1/ping")
-          val str = s.mkString
-          s.close()
-          str
-        }
-        assert(html.toLowerCase(Locale.ROOT).equals("pong"))
-      }
-    }
+    val resp = webTarget.path("/api/v1/ping").request().get()
+    resp.getStatusInfo
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/server/KyuubiRestFrontendServiceSuite.scala:26]: nice catch

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala [hunk 13]
```diff
@@ -17,116 +17,116 @@
...
-        statementHandleStr.replaceAll("EXECUTE_STATEMENT", "GET_TYPE_INFO")
-      response = webTarget.path(s"api/v1/operations/$invalidOperationHandle/event")
-        .request(MediaType.APPLICATION_JSON_TYPE).get()
-      assert(404 == response.getStatus)
-    }
+  test("get an operation event") {
+    val catalogsHandleStr = getOpHandleStr(OperationType.GET_CATALOGS)
+    var response = webTarget.path(s"api/v1/operations/$catalogsHandleStr/event")
+      .request(MediaType.APPLICATION_JSON_TYPE).get()
+    val operationEvent = response.readEntity(classOf[KyuubiOperationEvent])
```
Comment (cfmcgrady) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala:39]: never be called.

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala [hunk 14]
```diff
@@ -17,116 +17,116 @@
...
+    assert(200 == response.getStatus)
```
Comment (cfmcgrady) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala:40]: move to `checkOpState`?

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala [hunk 15]
```diff
@@ -17,116 +17,116 @@
...
+    checkOpState(catalogsHandleStr, FINISHED)
+
+    val statementHandleStr = getOpHandleStr(OperationType.EXECUTE_STATEMENT)
+    response = webTarget.path(s"api/v1/operations/$statementHandleStr/event")
+      .request(MediaType.APPLICATION_JSON_TYPE).get()
+    val statementEvent = response.readEntity(classOf[KyuubiOperationEvent])
```
Comment (cfmcgrady) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala:46]: ditto

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala [hunk 16]
```diff
@@ -34,315 +32,258 @@ import org.apache.kyuubi.session.SessionHandle
...
+
+    val getColumnsReq = GetColumnsRequest("spark_catalog", "default", "default", "default")
+    response = webTarget.path(s"$pathPrefix/operations/columns")
+      .request(MediaType.APPLICATION_JSON_TYPE)
+      .post(Entity.entity(getColumnsReq, MediaType.APPLICATION_JSON_TYPE))
+    assert(200 == response.getStatus)
+    operationHandle = response.readEntity(classOf[OperationHandle])
+    assert(operationHandle.typ == OperationType.GET_COLUMNS)
+
+    var getFunctionsReq = GetFunctionsRequest("default", "default", "default")
```
Comment (cfmcgrady) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala:242]: nit: `var -> val`

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala [hunk 17]
```diff
@@ -34,315 +32,258 @@ import org.apache.kyuubi.session.SessionHandle
...
+    val pathPrefix = s"api/v1/sessions/$serializedSessionHandle"
 
-      val serializedOperationHandle = s"${operationHandle.identifier.publicId}|" +
-        s"${operationHandle.identifier.secretId}|${operationHandle.protocol.getValue}|" +
-        s"${operationHandle.typ.toString}"
+    response = webTarget.path(s"$pathPrefix/operations/catalogs")
+      .request(MediaType.APPLICATION_JSON_TYPE)
+      .post(Entity.entity(null, MediaType.APPLICATION_JSON_TYPE))
+    assert(200 == response.getStatus)
+    var operationHandle = response.readEntity(classOf[OperationHandle])
```
Comment (cfmcgrady) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala:273]: nit: `var -> val`

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala (hunk 13 above)
Comment (yaooqinn) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala]: nice catch

**Review by cfmcgrady — APPROVED** [review]

File: kyuubi-server/src/test/scala/org/apache/kyuubi/RestFrontendTestHelper.scala [hunk 18]
```diff
@@ -46,40 +48,29 @@ object RestFrontendTestHelper {
...
-        webTarget)
-    } finally {
-      restApiBaseSuite.tearDown()
-      server.stop()
-    }
   }
+
+  override def afterAll(): Unit = {
+    restApiBaseSuite.tearDown()
+    server.stop()
```
Comment (pan3793) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/RestFrontendTestHelper.scala:67]: unnecessary `server.stop()`?

**Review by pan3793 — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**